File: coco/pattern/sensec/app/services/rotation_service.py
"""
KEK rotation service.

Strategy:
  1. Create a new KEK version in Azure Key Vault (or it was auto-rotated by AKV policy).
  2. Page through all EDEK records with status=current.
  3. For each: unwrap with old version → re-wrap with new version → update record.
  4. Mark old records as 'rotated' only after successful re-wrap.
  5. Old KEK version remains in AKV (disabled) so in-flight decrypts still work
     during the rotation window.
"""

from __future__ import annotations

import base64
from datetime import datetime, timezone

import structlog
from sqlalchemy import select, update
from sqlalchemy.ext.asyncio import AsyncSession, async_sessionmaker

from app.audit.logger import audit_log
from app.crypto.kek_client import KEKClient
from app.models.edek_record import EDEKRecord, RotationStatus
from app.models.schemas import RotateKEKResponse

log = structlog.get_logger("rotation_service")

PAGE_SIZE = 200


async def rotate_kek(
    kek_client: KEKClient,
    session_factory: async_sessionmaker,
    triggered_by: str = "scheduler",
) -> RotateKEKResponse:
    new_version = await kek_client.get_current_kek_version()
    log.info("kek_rotation_started", new_kek_version=new_version, triggered_by=triggered_by)

    total = 0
    offset = 0

    while True:
        async with session_factory() as session:
            rows = (
                await session.scalars(
                    select(EDEKRecord)
                    .where(EDEKRecord.rotation_status == RotationStatus.current)
                    .where(EDEKRecord.kek_version != new_version)
                    .order_by(EDEKRecord.created_at)
                    .limit(PAGE_SIZE)
                    .offset(offset)
                )
            ).all()

            if not rows:
                break

            for record in rows:
                await _rewrap_record(record, new_version, kek_client, session)
                total += 1

            await session.commit()
            offset += PAGE_SIZE

    audit_log(
        "kek_rotation_completed",
        new_kek_version=new_version,
        records_rotated=total,
        triggered_by=triggered_by,
        status="success",
    )

    return RotateKEKResponse(new_kek_version=new_version, records_queued=total)


async def _rewrap_record(
    record: EDEKRecord,
    new_version: str,
    kek_client: KEKClient,
    session: AsyncSession,
) -> None:
    old_edek = base64.b64decode(record.edek_blob)
    dek_bytes = await kek_client.unwrap_dek(old_edek, record.kek_version)
    new_edek, _ = await kek_client.wrap_dek(dek_bytes)

    record.edek_blob = base64.b64encode(new_edek).decode()
    record.kek_version = new_version
    record.rotation_status = RotationStatus.current
    record.rotated_at = datetime.now(timezone.utc)

File: coco/pattern/springsec/hsm-encryption-service/src/main/java/com/hsm/encryption/service/RotationService.java
package com.hsm.encryption.service;

import com.hsm.encryption.audit.AuditLogger;
import com.hsm.encryption.crypto.KekClient;
import com.hsm.encryption.dto.RotateKekResponse;
import com.hsm.encryption.model.EdekRecord;
import com.hsm.encryption.model.RotationStatus;
import com.hsm.encryption.repository.EdekRecordRepository;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.stereotype.Service;
import org.springframework.transaction.PlatformTransactionManager;
import org.springframework.transaction.support.TransactionTemplate;

import java.time.OffsetDateTime;
import java.util.Base64;
import java.util.List;

/**
 * KEK rotation service. Ported from app/services/rotation_service.py.
 *
 * <p>Strategy: page through all EDEK records with status=current whose kek_version
 * differs from the target version; for each, unwrap with the old version, re-wrap
 * with the new one, and update the record; commit per page. The old KEK version
 * stays usable in Key Vault (disabled, not deleted) so in-flight decrypts during
 * the rotation window still work.
 *
 * <p>Always re-queries page 0: once a page's records are re-wrapped their
 * kek_version equals the target, so they naturally drop out of the "current &amp;
 * kek_version != target" filter -- the next unprocessed batch is always at
 * offset 0. (The Python source instead increments a page offset across a
 * shrinking filtered result set, which would skip records; this port fixes that.)
 */
@Service
public class RotationService {

    private static final Logger log = LoggerFactory.getLogger(RotationService.class);
    private static final int PAGE_SIZE = 200;

    private final KekClient kekClient;
    private final EdekRecordRepository edekRecordRepository;
    private final AuditLogger auditLogger;
    private final TransactionTemplate transactionTemplate;

    public RotationService(KekClient kekClient, EdekRecordRepository edekRecordRepository,
                            AuditLogger auditLogger, PlatformTransactionManager transactionManager) {
        this.kekClient = kekClient;
        this.edekRecordRepository = edekRecordRepository;
        this.auditLogger = auditLogger;
        this.transactionTemplate = new TransactionTemplate(transactionManager);
    }

    public RotateKekResponse rotateKek(String triggeredBy) {
        String newVersion = kekClient.getCurrentKekVersion();
        log.info("kek_rotation_started new_kek_version={} triggered_by={}", newVersion, triggeredBy);

        int total = 0;
        while (true) {
            Page<EdekRecord> page = edekRecordRepository.findByRotationStatusAndKekVersionNotOrderByCreatedAtAsc(
                    RotationStatus.CURRENT, newVersion, PageRequest.of(0, PAGE_SIZE));
            List<EdekRecord> records = page.getContent();
            if (records.isEmpty()) {
                break;
            }
            transactionTemplate.executeWithoutResult(status -> {
                for (EdekRecord record : records) {
                    rewrapRecord(record, newVersion);
                    edekRecordRepository.save(record);
                }
            });
            total += records.size();
        }

        auditLogger.log("kek_rotation_completed",
                "new_kek_version", newVersion, "records_rotated", total, "triggered_by", triggeredBy, "status", "success");

        return new RotateKekResponse(newVersion, total);
    }

    private void rewrapRecord(EdekRecord record, String newVersion) {
        byte[] oldEdek = Base64.getDecoder().decode(record.getEdekBlob());
        byte[] dekBytes = kekClient.unwrapDek(oldEdek, record.getKekVersion());
        KekClient.WrapResult wrapResult = kekClient.wrapDek(dekBytes);

        record.setEdekBlob(Base64.getEncoder().encodeToString(wrapResult.edekBytes()));
        record.setKekVersion(newVersion);
        record.setRotationStatus(RotationStatus.CURRENT);
        record.setRotatedAt(OffsetDateTime.now());
    }
}
